cpp/src/arrow/dataset/dataset_writer_test.cc

```diff
@@ -265,6 +267,14 @@ TEST_F(DatasetWriterTestFixture, DirectoryCreateFails) {
   ASSERT_FINISHES_AND_RAISES(Invalid, scheduler_finished_);
 }
 
+TEST_F(DatasetWriterTestFixture, BatchGreaterThanMaxRowsQueued) {
```

Comment:
Sorry I can't see why the deadlock happens in this case. Do you mind to elaborate a bit? Thanks.